File: autonomy/camera-radar-fusion/src/shared/fusion/services/visualize_camera_radar_fusion.py
import logging
from pathlib import Path

import cv2
import numpy as np
from config import config
from nuscenes.nuscenes import NuScenes
from scipy.spatial.transform import Rotation
from shared.constants.colors import COLOR_BLACK, COLOR_BLUE, COLOR_GREEN, COLOR_YELLOW
from shared.fusion.services.fuse_camera_radar import fuse_camera_radar
from shared.fusion.utils.convert_nuscenes_quaternion_to_scipy import (
    convert_nuscenes_quaternion_to_scipy,
)
from shared.fusion.utils.project_radar_to_camera import project_radar_to_camera
from shared.radar.services.load_radar_data import load_radar_data
from ultralytics import YOLO

logger = logging.getLogger(__name__)


def visualize_camera_radar_fusion(
    nuscenes_instance: NuScenes,
    nuscenes_scene_index: int,
    visualization_frame_count: int,
) -> None:
    """
    Visualize camera-radar fusion for a nuScenes scene.

    Args:
        nuscenes_instance: NuScenes dataset instance
        nuscenes_scene_index: Index of scene to visualize (0-9 for mini dataset)
        visualization_frame_count: Maximum number of frames to process

    """
    # Load YOLO model for camera object detection
    logger.info("Loading YOLO model...")
    yolo_model = YOLO(Path(config.YOLO_MODEL_PATH))
    logger.info("YOLO model loaded")

    # Get scene
    scene = nuscenes_instance.scene[nuscenes_scene_index]
    logger.info(
        f"Processing scene: {scene['name']}, Description: {scene['description']}",
    )

    # Get first sample
    sample_token = scene["first_sample_token"]

    frame_count = 0
    while sample_token != "" and frame_count < visualization_frame_count:
        sample = nuscenes_instance.get("sample", sample_token)

        # Get front camera data
        camera_token = sample["data"]["CAM_FRONT"]
        camera_sample = nuscenes_instance.get("sample_data", camera_token)
        camera_path = nuscenes_instance.get_sample_data_path(camera_token)

        # Load camera image
        image = cv2.imread(str(camera_path))
        if image is None:
            logger.warning(f"Failed to load image: {camera_path}")
            sample_token = sample["next"]
            continue

        # Get camera calibration
        camera_calibration_token = camera_sample["calibrated_sensor_token"]
        camera_calibration = nuscenes_instance.get(
            "calibrated_sensor",
            camera_calibration_token,
        )
        camera_intrinsic = np.array(camera_calibration["camera_intrinsic"])

        # Get front radar data (RADAR_FRONT)
        radar_token = sample["data"]["RADAR_FRONT"]
        radar_sample = nuscenes_instance.get("sample_data", radar_token)

        # Load radar data
        radar_data = load_radar_data(nuscenes_instance, radar_token)

        # Get transformations
        radar_calibration_token = radar_sample["calibrated_sensor_token"]
        radar_calibration = nuscenes_instance.get(
            "calibrated_sensor",
            radar_calibration_token,
        )

        # Build transformation matrix from radar to camera
        # First: radar -> vehicle, then: vehicle -> camera
        radar_to_vehicle = np.eye(4)
        radar_quaternion_scipy = convert_nuscenes_quaternion_to_scipy(
            radar_calibration["rotation"],
        )
        radar_rotation = Rotation.from_quat(radar_quaternion_scipy)
        radar_to_vehicle[:3, :3] = radar_rotation.as_matrix()
        radar_to_vehicle[:3, 3] = np.array(radar_calibration["translation"])

        # Build camera to vehicle transformation (then invert to get vehicle to camera)
        camera_to_vehicle = np.eye(4)
        camera_quaternion_scipy = convert_nuscenes_quaternion_to_scipy(
            camera_calibration["rotation"],
        )
        camera_rotation = Rotation.from_quat(camera_quaternion_scipy)
        camera_to_vehicle[:3, :3] = camera_rotation.as_matrix()
        camera_to_vehicle[:3, 3] = np.array(camera_calibration["translation"])

        # Invert to get vehicle to camera
        vehicle_to_camera = np.linalg.inv(camera_to_vehicle)

        radar_to_camera = vehicle_to_camera @ radar_to_vehicle

        # Project radar points to camera
        radar_points_3d = radar_data[:, :3]
        radar_velocities = radar_data[:, 3]
        radar_cross_sections = radar_data[:, 4]

        image_points = project_radar_to_camera(
            radar_points_3d,
            camera_intrinsic,
            radar_to_camera,
        )

        # Perform sensor fusion
        fused_tracks, unmatched_camera, unmatched_radar = fuse_camera_radar(
            image=image,
            radar_points_3d=radar_points_3d,
            radar_velocities=radar_velocities,
            radar_cross_sections=radar_cross_sections,
            image_points=image_points,
            yolo_model=yolo_model,
        )

        # Visualize
        visualization = image.copy()

        # Draw fused tracks (green bounding boxes with radar info)
        for track in fused_tracks:
            x1, y1, x2, y2 = track.bounding_box.astype(int)

            # Bounding box for fused tracks
            cv2.rectangle(visualization, (x1, y1), (x2, y2), COLOR_GREEN, 2)

            # Label with class, distance, and velocity
            label = f"{track.class_name} {track.distance:.1f}m"
            if track.is_moving:
                label += f" {track.velocity:.1f}m/s"

            # Background for text
            text_size = cv2.getTextSize(
                label,
                cv2.FONT_HERSHEY_SIMPLEX,
                0.5,
                1,
            )[0]
            cv2.rectangle(
                visualization,
                (x1, y1 - text_size[1] - 4),
                (x1 + text_size[0], y1),
                COLOR_GREEN,
                -1,
            )

            # Label text
            cv2.putText(
                visualization,
                label,
                (x1, y1 - 2),
                cv2.FONT_HERSHEY_SIMPLEX,
                0.5,
                COLOR_BLACK,
                1,
            )

            # Draw center point
            center_x = int(track.image_coordinate_x)
            center_y = int(track.image_coordinate_y)
            cv2.circle(visualization, (center_x, center_y), 5, COLOR_GREEN, -1)

        # Draw unmatched camera detections (blue bounding boxes)
        for camera_detection in unmatched_camera:
            x1, y1, x2, y2 = camera_detection.bounding_box.astype(int)

            # Blue bounding box for camera-only detections
            cv2.rectangle(visualization, (x1, y1), (x2, y2), COLOR_BLUE, 2)

            label = f"{camera_detection.class_name} (cam only)"

            # Label text
            cv2.putText(
                visualization,
                label,
                (x1, y1 - 5),
                cv2.FONT_HERSHEY_SIMPLEX,
                0.4,
                COLOR_BLUE,
                1,
            )

        # Draw unmatched radar detections (yellow points)
        for radar_detection in unmatched_radar:
            pixel_x = int(radar_detection.image_coordinate_x)
            pixel_y = int(radar_detection.image_coordinate_y)

            # Check if point is within image bounds
            if (
                0 <= pixel_x < visualization.shape[1]
                and 0 <= pixel_y < visualization.shape[0]
            ):
                # Yellow circle for radar-only detections
                radius = max(
                    3,
                    min(10, int(radar_detection.radar_cross_section / 10)),
                )
                cv2.circle(visualization, (pixel_x, pixel_y), radius, COLOR_YELLOW, -1)

                # Draw distance
                cv2.putText(
                    visualization,
                    f"{radar_detection.distance:.1f}m",
                    (pixel_x + 8, pixel_y - 8),
                    cv2.FONT_HERSHEY_SIMPLEX,
                    0.4,
                    COLOR_YELLOW,
                    1,
                )

        # Add info overlay
        info_y = 30
        cv2.putText(
            visualization,
            f"Scene: {scene['name']} | Frame: {frame_count + 1}",
            (10, info_y),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.7,
            (255, 255, 255),
            2,
        )
        info_y += 30
        cv2.putText(
            visualization,
            f"Fused: {len(fused_tracks)} | Camera-only: {len(unmatched_camera)} | "
            f"Radar-only: {len(unmatched_radar)}",
            (10, info_y),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.6,
            (255, 255, 255),
            2,
        )

        # Legend
        legend_y = visualization.shape[0] - 80
        cv2.putText(
            visualization,
            "Green Box = Fused Track (Camera + Radar)",
            (10, legend_y),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.5,
            (0, 255, 0),
            1,
        )
        legend_y += 20
        cv2.putText(
            visualization,
            "Blue Box = Camera Only | Yellow Point = Radar Only",
            (10, legend_y),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.5,
            (255, 255, 255),
            1,
        )

        # Show image
        cv2.imshow("Camera-Radar Fusion", visualization)

        key = cv2.waitKey(100) & 0xFF  # 10 FPS playback
        if key == ord("q"):
            break
        if key == ord(" "):
            # Pause on space
            cv2.waitKey(0)

        # Move to next sample
        sample_token = sample["next"]
        frame_count += 1

    cv2.destroyAllWindows()
    logger.info(f"Processed {frame_count} frames from scene {scene['name']}")

File: autonomy/camera-radar-lidar-fusion/src/shared/fusion/services/visualize_camera_radar_fusion.rs
use crate::config::AppConfig;
use crate::shared::camera::services::detect_objects_in_camera::{
    YoloModel, detect_objects_in_camera,
};
use crate::shared::fusion::constants::colors::{
    COLOR_BLACK_SCALAR, COLOR_BLUE_SCALAR, COLOR_YELLOW_SCALAR,
};
use crate::shared::fusion::services::fuse_camera_radar::fuse_camera_radar;
use crate::shared::fusion::utils::calculate_distance_color::calculate_distance_color;
use crate::shared::radar::services::create_radar_detection::create_radar_detection;
use crate::shared::radar::services::load_radar_data::load_radar_data;
use crate::shared::radar::utils::project_radar_to_camera::project_radar_to_camera;
use crate::shared::rerun::constants::entity_paths::FUSION_VISUALIZATION_ENTITY_PATH;
use crate::shared::rerun::services::log_rerun_image::log_rerun_image;
use anyhow::{Context, Result};
use nalgebra::{Matrix3, Matrix4, Vector3};
use opencv::core::{Point, Rect, Scalar};
use opencv::imgcodecs::imread;
use opencv::imgproc::{HersheyFonts, LINE_8, get_text_size, put_text, rectangle};
use opencv::prelude::MatTraitConst;
use rerun as rr;
use std::path::Path;
use tracing::warn;

pub fn visualize_camera_radar_fusion<P: AsRef<Path>>(
    recording: &rr::RecordingStream,
    camera_image_path: P,
    radar_data_path: P,
    camera_intrinsic: Matrix3<f64>,
    radar_to_camera: Matrix4<f64>,
    yolo_model: &mut YoloModel,
    config: AppConfig,
) -> Result<()> {
    let image = imread(
        camera_image_path
            .as_ref()
            .to_str()
            .expect("Failed to convert camera image path to string"),
        opencv::imgcodecs::IMREAD_COLOR,
    )
    .context("Failed to load camera image")?;

    if image.empty() {
        warn!("Failed to load image");
        return Ok(());
    }

    let radar_data = load_radar_data(radar_data_path)?;

    let radar_points_3d = radar_data.fixed_rows::<3>(0);
    let radar_velocities = radar_data.row(3);
    let radar_cross_sections = radar_data.row(4);

    let radar_points_vec: Vec<Vector3<f64>> = (0..radar_points_3d.ncols())
        .map(|column_index| {
            Vector3::new(
                radar_points_3d[(0, column_index)] as f64,
                radar_points_3d[(1, column_index)] as f64,
                radar_points_3d[(2, column_index)] as f64,
            )
        })
        .collect();

    let image_points =
        project_radar_to_camera(&radar_points_vec, &camera_intrinsic, &radar_to_camera);

    if image_points.is_empty() {
        warn!("No radar points projected into camera frame");
    }

    let camera_detections = detect_objects_in_camera(&image, yolo_model, 0.6)?;

    let mut radar_detections = Vec::new();
    for index in 0..radar_points_3d.ncols() {
        if index >= image_points.len() {
            warn!(
                "Projection list shorter than radar points; skipping index {}",
                index
            );
            break;
        }
        let projected = match image_points[index] {
            Some(value) => value,
            None => continue,
        };
        let point_3d = Vector3::new(
            radar_points_3d[(0, index)] as f64,
            radar_points_3d[(1, index)] as f64,
            radar_points_3d[(2, index)] as f64,
        );
        let velocity = radar_velocities[index] as f64;
        let radar_cross_section = radar_cross_sections[index] as f64;
        let image_x = projected[0];
        let image_y = projected[1];

        radar_detections.push(create_radar_detection(
            point_3d,
            velocity,
            radar_cross_section,
            image_x,
            image_y,
        ));
    }

    let fusion_result = fuse_camera_radar(
        camera_detections,
        radar_detections,
        config.association_distance_threshold_pixels,
        &config,
    );

    let mut visualization = image.clone();

    for track in &fusion_result.fused_tracks {
        let bounding_box = track.bounding_box();
        let x1 = bounding_box[0] as i32;
        let y1 = bounding_box[1] as i32;
        let x2 = bounding_box[2] as i32;
        let y2 = bounding_box[3] as i32;

        rectangle(
            &mut visualization,
            Rect::new(x1, y1, x2 - x1, y2 - y1),
            COLOR_YELLOW_SCALAR,
            2,
            LINE_8,
            0,
        )?;

        let label = if track.is_moving(&config) {
            format!(
                "{} {:.1}m {:.1}m/s",
                track.class_name(),
                track.distance(),
                track.velocity()
            )
        } else {
            format!("{} {:.1}m", track.class_name(), track.distance())
        };

        let text_size = get_text_size(
            &label,
            HersheyFonts::FONT_HERSHEY_SIMPLEX as i32,
            0.5,
            1,
            &mut 0,
        )?;

        rectangle(
            &mut visualization,
            Rect::new(
                x1,
                y1 - text_size.height - 4,
                text_size.width,
                text_size.height + 4,
            ),
            COLOR_YELLOW_SCALAR,
            -1,
            LINE_8,
            0,
        )?;

        put_text(
            &mut visualization,
            &label,
            Point::new(x1, y1 - 2),
            HersheyFonts::FONT_HERSHEY_SIMPLEX as i32,
            0.5,
            COLOR_BLACK_SCALAR,
            1,
            LINE_8,
            false,
        )?;

        let center_x = track.image_coordinate_x() as i32;
        let center_y = track.image_coordinate_y() as i32;
        opencv::imgproc::circle(
            &mut visualization,
            Point::new(center_x, center_y),
            5,
            COLOR_YELLOW_SCALAR,
            -1,
            LINE_8,
            0,
        )?;
    }

    for camera_detection in &fusion_result.unmatched_camera_detections {
        let bounding_box = camera_detection.bounding_box;
        let x1 = bounding_box[0] as i32;
        let y1 = bounding_box[1] as i32;
        let x2 = bounding_box[2] as i32;
        let y2 = bounding_box[3] as i32;

        rectangle(
            &mut visualization,
            Rect::new(x1, y1, x2 - x1, y2 - y1),
            COLOR_BLUE_SCALAR,
            2,
            LINE_8,
            0,
        )?;

        let label = format!("{} (cam only)", camera_detection.class_name);

        put_text(
            &mut visualization,
            &label,
            Point::new(x1, y1 - 5),
            HersheyFonts::FONT_HERSHEY_SIMPLEX as i32,
            0.4,
            COLOR_BLUE_SCALAR,
            1,
            LINE_8,
            false,
        )?;
    }

    for radar_detection in &fusion_result.unmatched_radar_detections {
        let pixel_x = radar_detection.image_coordinate_x as i32;
        let pixel_y = radar_detection.image_coordinate_y as i32;

        if pixel_x >= 0
            && pixel_x < visualization.cols()
            && pixel_y >= 0
            && pixel_y < visualization.rows()
        {
            let color = calculate_distance_color(
                radar_detection.distance,
                config.association_distance_threshold_pixels,
            );
            let radius = ((radar_detection.radar_cross_section / 10.0) as i32).clamp(3, 10);
            opencv::imgproc::circle(
                &mut visualization,
                Point::new(pixel_x, pixel_y),
                radius,
                color,
                -1,
                LINE_8,
                0,
            )?;

            put_text(
                &mut visualization,
                &format!("{:.1}m", radar_detection.distance),
                Point::new(pixel_x + 8, pixel_y - 8),
                HersheyFonts::FONT_HERSHEY_SIMPLEX as i32,
                0.4,
                color,
                1,
                LINE_8,
                false,
            )?;
        }
    }

    let info_text = format!(
        "Fused: {} | Camera-only: {} | Radar-only: {}",
        fusion_result.fused_tracks.len(),
        fusion_result.unmatched_camera_detections.len(),
        fusion_result.unmatched_radar_detections.len()
    );
    put_text(
        &mut visualization,
        &info_text,
        Point::new(10, 30),
        HersheyFonts::FONT_HERSHEY_SIMPLEX as i32,
        0.6,
        Scalar::new(255.0, 255.0, 255.0, 0.0),
        2,
        LINE_8,
        false,
    )?;

    let legend_text = "Yellow Box = Fused Track (Camera + Radar) | Blue Box = Camera Only | Radar Point Color: Red (Near) -> Yellow (Mid) -> Green (Far)";
    put_text(
        &mut visualization,
        legend_text,
        Point::new(10, 60),
        HersheyFonts::FONT_HERSHEY_SIMPLEX as i32,
        0.5,
        Scalar::new(255.0, 255.0, 255.0, 0.0),
        1,
        LINE_8,
        false,
    )?;

    log_rerun_image(recording, &visualization, FUSION_VISUALIZATION_ENTITY_PATH)?;

    Ok(())
}
